### PR TITLE
:broom: clean up and upgrade python syntax.

### PR DESCRIPTION
**Pull request by Harmouch101**
Signed-off-by: Harmouch101 <eng.mahmoudharmouch@gmail.com>

### What was wrong?

Related to Issue #

### How was it fixed?
As of python3, 'Unicode' strings are the default. Hence no need for the prefix `u`. Similarly, `__nonzero__` is a python2 syntax for defining boolness. I also used f-strings since this project support 3.6+.

So, I manually went through the codebase, updated:

- [x] old-school string formatting to f-strings.
- [x] list to set whenever possible ~~because sets are more efficient.~~ to avoid redundant calls of `set`.
- [x] Some other cleanups. 
- [x] ~~Some of the error messages:~~

```python
        message = (
            f"\nCould not identify the intended function with name `"
            f"{fn_identifier}`, positional argument"
            f"{'s' if len(args) > 1 else ''} of type "
            f"`{tuple(map(type, args))}` and keyword argument"
            f"{'s' if len(kwargs) > 1 else ''} of type "
            f"`{valmap(type, kwargs)}`.\nFound {len(matching_identifiers)} "
            f"function{'s' if len(matching_identifiers) > 1 else ''} with "
            f"the name `{fn_identifier}`"
            f": {matching_function_signatures}{diagnosis}"
        )
```

output example:

```sh
"\nCould not identify the intended function with name `setBytes32Value`, positional argument of type `(<class 'str'>,)` and keyword argument of type `{}`.\nFound 1 function with the name `setBytes32Value`: ['setBytes32Value(bytes32[])']\nFunction invocation failed due to no matching argument types.".
```

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)

#### Cute Animal Picture

:dog:

### CODE REVIEW

**Comment by Harmouch101** [comment]
Apparently, f-strings seem to be the fastest string formatting method in microbenchmarks as [Raymond Hettinger](https://twitter.com/raymondh/status/1205969258800275456) showed.

**Comment by Harmouch101** [comment]
I think this PR is ready for reviews from the warriors @fselmo, @kclowes, @pipermerriam ...

File: tests/ethpm/validation/test_manifest.py
Comment (fselmo) [inline tests/ethpm/validation/test_manifest.py:91]: Alternatively, I _think_ that we can keep `actual == expected` and avoid calling `set()` on the expected sets by leaving `set()` instead of `{}` in line `81`:

```py
@pytest.mark.parametrize(
    "data,expected",
    (
        ([], set()),
        ([{"some": {"contractType": "one"}}], {"one"}),
        (
            [{"some": {"contractType": "one"}, "other": {"contractType": "two"}}],
            {"one", "two"},
        ),
    ),
)
def test_extract_contract_types_from_deployments(data, expected):
    actual = extract_contract_types_from_deployments(data)
    assert actual == expected
```

**Review by fselmo — COMMENTED** [review]
Hey @Harmouch101. Wow. Thanks for the hefty sweeping! I think in the future it might be nicer to break the different changes up into separate PRs for large loads like this but bravo 👍 and thank you 🙏 . We'd been steering towards f-strings in general, changing it where it made sense, and this certainly helps us get there faster 😅.

I left some nit comments where it's nice that f-strings are shorter since we can fit messages in one line rather than line split.

I think I found just one misrepresentation / typo from the old `.format()` to `f-string` format that I didn't preface with a "**nit:**" tag.

I left a comment on the pluralization logic for strings in general. I think that though the pluralization being accurate is nice, I'm not sure that outweighs the same logic being repeated across all strings in the code base where things might need to be plural. I think it might be better to have static messages in general too with just the expected variables being dynamic - for consistency to the end-user.

I could've totally overlooked something too so let me know. Thanks again though for another very thorough PR.

---

**edit:** I forgot to mention it but I'm a big fan of the `list` -> `set` changes where it made sense 👌

**Review by kclowes — COMMENTED** [review]
LGTM after the comments from @fselmo are addressed! Thanks @Harmouch101!

File: web3/_utils/caching.py
Comment (Harmouch101) [reply web3/_utils/caching.py:42]: Actually, my editor automatically reformats the line because I set the max length of the line to be 70, I forgot that in this project it is set to 100 chars.

File: web3/_utils/blocks.py
Comment (Harmouch101) [reply web3/_utils/blocks.py:73]: Agree! I did the second one!

File: web3/_utils/abi.py
Comment (Harmouch101) [reply web3/_utils/abi.py]: I think for better code readability we can do something like:

```
    if duplicate_args:
        s = 's' if len(duplicate_args) > 1 else ''
        duplicate_args = ', '.join(duplicate_args)
        raise TypeError(
            f"{function_abi.get('name')}() got multiple values for argument{s} '{duplicate_args}'"
        )
```

File: web3/_utils/abi.py
Comment (Harmouch101) [reply web3/_utils/abi.py]: Yup!

File: web3/_utils/empty.py
Comment (Harmouch101) [reply web3/_utils/empty.py:10]: Thank you!

File: web3/_utils/encoding.py
Comment (Harmouch101) [reply web3/_utils/encoding.py]: Roger!

File: web3/main.py
Comment (Harmouch101) [reply web3/main.py]: Yup!

File: web3/main.py
Comment (Harmouch101) [reply web3/main.py]: Yes, but without the last two `f` prefixes:
 
```python
raise TypeError(
    f"You called keccak with first arg {primitive!r} and keywords {{'text': {text!r}, "
    f"'hexstr': {hexstr!r}}}. You must call it with one of these approaches: "
    "keccak(text='txt'), keccak(hexstr='0x747874'), keccak(b'\\x74\\x78\\x74'), "
    "or keccak(0x747874)."
)
```

File: web3/providers/auto.py
Comment (Harmouch101) [reply web3/providers/auto.py]: Copy!

File: web3/providers/eth_tester/defaults.py
Comment (Harmouch101) [reply web3/providers/eth_tester/defaults.py]: Roger that!

File: web3/providers/ipc.py
Comment (Harmouch101) [reply web3/providers/ipc.py]: That's right!

File: web3/providers/auto.py
Comment (Harmouch101) [reply web3/providers/auto.py]: yup!

File: tests/ethpm/validation/test_manifest.py
Comment (Harmouch101) [reply tests/ethpm/validation/test_manifest.py]: :+1:

File: web3/_utils/abi.py
Comment (fselmo) [reply web3/_utils/abi.py]: Well we talked it through too and static messages (predictable text) is better practice, with only values / variables being dynamic in the message. It can also be a source of bugs we have to go through and fix since it's not tested, etc. This seems like the only change left. I wouldn't mind making the change though if you want and merging it. Let me know!

Comment (Harmouch101) [reply web3/_utils/abi.py]: Alright. Well, i don't want to bother you with this issue, but if you have time, you are good to go. At the same time i can do it myself, cause there are other strings i have changed that look like this. And oh, forgot to mention there is a test case that test these error messages which i have altered to fit this schema.
If you are willing to make the changes, i would suggest to search for ` = 's' if len(` across the whole code base, and make the changes accordingly.
Right now i am AFK, texting you using the app.

Regards.

**Comment by Harmouch101** [comment]
@fselmo, Just ping to let you know I have done the changes requested.

**Comment by fselmo** [comment]
> @fselmo, Just ping to let you know I have done the changes requested.

Just getting back to this. Thanks again for these changes @Harmouch101 👌

**Review by fselmo — APPROVED** [review]

**Comment by Harmouch101** [comment]
Hey @fselmo , just a suggestion about the commit messages. I think for the future it would be better to squash all commits into a single one cause some of my commit messages doesn't make sense for the whole code base, rather for this specific PR. Therefore, i suggest that the contribution file should mention that the authors of PRs must squash their commits when they finish pushing all the necessary code for a specific PR. Or, actually, to make life easier for new contributors, the task can be done by maintainers.

Regards.

**Comment by fselmo** [comment]
Hey @Harmouch101. Agreed. We usually do squash commits. I overlooked it before merging this time.

**Pull request merged**
